build_rules.gradle

```diff
@@ -0,0 +1,275 @@
+  jcenter()
+
+  // Release staging repository
+  maven { url "https://oss.sonatype.org/content/repositories/staging/" }
+
+  // Apache nightly snapshots
+  maven { url "https://repository.apache.org/snapshots" }
+
+  // Apache release snapshots
+  maven { url "https://repository.apache.org/content/repositories/releases" }
+}
+
+// Provide code coverage
+apply plugin: "jacoco"
+
+// Provies "htmlDependencyReport" task which spits out a nice page per subproject
+// showing all the dependencies per Gradle configuration.
+apply plugin: "project-report"
+
+// Apply a task dependency visualization plugin which creates a ".dot" file in the build directory
+// giving the task dependencies for that project
+apply plugin: "cz.malohlava.visteg"
+
+/*************************************************************************************************/
+
+/** Used as configuration within the applyJavaNature closure. */
+class JavaNatureConfiguration {
+  double javaVersion = 1.7
+  boolean enableFindbugs = true
+  boolean enableShadow = true
+}
+
+/** Takes a map of parameters. See JavaNatureConfiguration for the set of parameters allowed. */
+ext.applyJavaNature = {
+  println "applyJavaNature with " + (it ? "$it" : "default configuration") + " for project $project.name"
+  // Use the implicit it parameter of the closure to handle zero argument or one argument map calls.
+  JavaNatureConfiguration configuration = it ? it as JavaNatureConfiguration : new JavaNatureConfiguration()
+  apply plugin: "maven"
+  apply plugin: "java"
+
+  sourceCompatibility = configuration.javaVersion
+  targetCompatibility = configuration.javaVersion
+  tasks.withType(JavaCompile) {
+    options.encoding = "UTF-8"
+  }
+
+  test {
+    // Configure the gradle include defaults to be equivalent to the maven-surefire-plugin
+    // http://maven.apache.org/components/surefire/maven-surefire-plugin/test-mojo.html
+    include "**/Test*.class"
+    include "**/*Test.class"
+    include "**/*Tests.class"
+    include "**/*TestCase.class"
+    useJUnit { }
+  }
+
+  apply plugin: "net.ltgt.apt"
+  dependencies {
+    def auto_value = "com.google.auto.value:auto-value:1.5.1"
+    def auto_service = "com.google.auto.service:auto-service:1.0-rc2"
+    compileOnly auto_value
+    apt auto_value
+    testCompileOnly auto_value
+    testApt auto_value
+
+    compileOnly auto_service
+    apt auto_service
+    testCompileOnly auto_service
+    testApt auto_service
+  }
+
+  // Add the optional and provided scopes for dependencies
+  apply plugin: 'propdeps'
+  apply plugin: 'propdeps-maven'
+
+  apply plugin: 'checkstyle'
+
+  tasks.withType(Checkstyle) {
+    configFile = project(":").file("sdks/java/build-tools/src/main/resources/beam/checkstyle.xml")
+    configProperties = [ "checkstyle.suppressions.file": project(":").file("sdks/java/build-tools/src/main/resources/beam/suppressions.xml") ]
+    showViolations = true
+    maxErrors = 0
+  }
+
+  apply plugin: 'com.diffplug.gradle.spotless'
```

Comment:
<!--new_thread; commit:8ca3fdca276a74973cd428d7321b995f7b7ebccf; resolved:0-->
Here, too, comments about what plugin is being loaded, why, how it is configured.